engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/TailCallMini.scala

```diff
@@ -0,0 +1,220 @@
+package org.enso.compiler.pass.analyse
+
+import org.enso.compiler.core.Implicits.{AsDiagnostics, AsMetadata}
+import org.enso.compiler.core.ir.MetadataStorage._
+import org.enso.compiler.core.ir.expression.{
+  Application,
+  Case,
+  Comment,
+  Foreign
+}
+import org.enso.compiler.core.ir.module.scope.{definition, Definition}
+import org.enso.compiler.core.ir.{
+  CallArgument,
+  DefinitionArgument,
+  Empty,
+  Expression,
+  Function,
+  Name,
+  Pattern,
+  Type,
+  Warning
+}
+import org.enso.compiler.core.{CompilerError, IR}
+import org.enso.compiler.pass.MiniIRPass
+import org.enso.compiler.pass.analyse.TailCall.TailPosition
+import org.enso.compiler.pass.resolve.ExpressionAnnotations
+import org.graalvm.collections.{EconomicSet, Equivalence}
+
+/** @param isInTailPosition Indicates whether the current IR element should be marked as tail call.
+  *                         This flag is propagated to the children during the preparation phase.
+  * @param tails A set of IR elements (collected and used only during the preparation phase)
+  *                         that are definitely tail calls, even if `isInTailPosition` is false.
+  *              EconomitSet with IDENTITY_SYSTEM_HASHCODE is chosen to the comparison of
+  *              references rather than via `IR.equals` and `IR.hashCode`.
+  * @param notTails A set of IR elements (collected and used only during the preparation phase)
+  *                 that are definitely not tail calls, even if `isInTailPosition` is true.
+  */
+class TailCallMini(
+  private val isInTailPosition: Boolean = false,
+  private val tails: EconomicSet[IR] =
+    EconomicSet.create(Equivalence.IDENTITY_WITH_SYSTEM_HASHCODE),
```

Comment:
I'd rather avoid dependency on `org.graalvm.collections` as introduced by 079543afd632f11da1a12f53585ca14e064f0dbd. Just makes the set of `runtime-compiler` dependencies needlessly heavyweight. Try:
```scala
private val tailsMap = new java.util.IdentityHashMap[IR, TailCallMini]();
```
and then use `put(_,this)` instead of `add` and `containsKey(...)` instead of `contains`.

Although the whole concept of computing everything in `prepare` and using it later kind of beats the purpose of `MiniIRPass`!

Reply:
Thanks for the reference. Replaced in edec03ba539132762a19b89df942d7adf35fc342